tests/test_utils_legacy.py

```diff
@@ -0,0 +1,77 @@
+from typing import Dict, List, Mapping, Optional, Sequence, Set, Tuple, Union
+
+from ollama._utils import _get_json_type, convert_function_to_tool
+
+
+def test_json_type_conversion():
+  # Test basic types
+  assert _get_json_type(str) == 'string'
+  assert _get_json_type(int) == 'integer'
+  assert _get_json_type(List) == 'array'
+  assert _get_json_type(Dict) == 'object'
+
+  # Test Optional
+  assert _get_json_type(Optional[str]) == 'string'
+
+
+def test_function_to_tool_conversion():
+  def add_numbers(x: int, y: Union[int, None] = None) -> int:  # Changed Optional to Union
+    """Add two numbers together.
+    Args:
+        x (integer): The first number
+        y (integer, optional): The second number
+
+    Returns:
+        integer: The sum of x and y
+    """
+    return x + y
+
+  tool = convert_function_to_tool(add_numbers)
+
+  assert tool['type'] == 'function'
+  assert tool['function']['name'] == 'add_numbers'
+  assert tool['function']['description'] == 'Add two numbers together.'
+  assert tool['function']['parameters']['type'] == 'object'
+  assert tool['function']['parameters']['properties']['x']['type'] == 'integer'
+  assert tool['function']['parameters']['properties']['x']['description'] == 'The first number'
+  assert tool['function']['parameters']['required'] == ['x']
+
+
+def test_function_with_all_typing_types():
+  def all_types(
+    x: int,
+    y: str,
+    z: Sequence[int],
+    w: Mapping[str, int],
+    d: Dict[str, int],
+    s: Set[int],
+    t: Tuple[int, str],
+    l: List[int],  # noqa: E741
+    o: Union[int, None],  # Changed Optional to Union
+  ) -> Union[Mapping[str, int], str, None]:
+    """
```

Comment:
Something important to note here is that if  `from __future__ import annotations` is used the parser breaks as it is currently parsing types directly and not a string. 

Specifically, it breaks for nested types like Union[..., ...] and Mapping[..., ...]. 

I could write a string parser if we think this is a big deal but it will be a bit of a time sink for sure. Good to decide what the support level should be python 3.9 and how opinionated we need to be. I think in this case particular the time sink is not worth the support but open to hearing thoughts.